server/src/Core.hs

```diff
@@ -139,40 +126,36 @@ enqueueCommand cmd core = atomically $ writeTBQueue (coreQueue core) cmd
 
 getCurrentValue :: Core -> Path -> IO (Maybe Value)
 getCurrentValue core path =
-  fmap (Store.lookup path) $ atomically $ readTVar $ coreCurrentValue core
+  fmap (Store.lookup path) $ atomically $ Persistence.getValue $ coreCurrentValue core
 
 withCoreMetrics :: MonadIO m => Core -> (Metrics.IcepeakMetrics -> IO ()) -> m ()
 withCoreMetrics core act = liftIO $ forM_ (coreMetrics core) act
 
--- Execute a modification.
-applyModification :: Modification -> Value -> Value
-applyModification (Delete path) value = Store.delete path value
-applyModification (Put path newValue) value = Store.insert path newValue value
-
 -- | Drain the command queue and execute them. Changes are published to all
 -- subscribers. This function returns when executing the 'Stop' command from the
 -- queue.
-runCommandLoop :: Core -> IO Value
-runCommandLoop core = atomically (readTVar (coreCurrentValue core)) >>= go
+runCommandLoop :: Core -> IO ()
+runCommandLoop core = go
   where
-    go val = do
+    go = do
       command <- atomically $ readTBQueue (coreQueue core)
       case command of
         Modify op -> do
-          let newValue = applyModification op val
-          atomically $ do
-            writeTVar (coreCurrentValue core) newValue
-            writeTVar (coreValueIsDirty core) True
-            writeTBQueue (coreUpdates core) (Just $ Updated (modificationPath op) newValue)
-          when (isNothing $ configSyncIntervalMicroSeconds $ coreConfig core) $
-            persistData core
-          go newValue
+          Persistence.apply op (coreCurrentValue core)
+          postUpdate (Store.modificationPath op) core
+          when (not $ periodicSyncingEnabled $ coreConfig core) $
```

Comment:
This reads a little confusing: If periodic syncing is not enabled then call sync.
Perhaps add a comment to point out that our default is to sync everything after every modification if periodic syncing is not enabled.

Nit: Instead of `when not` you could use `unless`.

Reply:
Explained